### ISSUE TITLE
Is there a script for offline alignment and full fine-tune? And what about catastrophic forgetting？

### ISSUE THREAD

**Issue by Labmem009** [wwxu21/CUT#1]
Interesting work!
Is there a script for offline alignment and full fine-tune?
And I wonder if the online alignment would cause catastrophic forgetting？
Thanks a lot!

**Comment by wwxu21** [comment]
Regarding the offline alignment:
Here is an example of fine-tuning the Shepherd dataset. (Just modify the ` --data-path` to be the offline dataset)

First, get the Shepherd dataset according to Sec. 2.1. 
Then use the following script:
```bash
threshold=1.2
weight_unlike=0.5
name=cut-1plus-13b
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 torchrun --nproc_per_node=8 --master_port=1233 finetune_unlikelihood.py \
    --base_model saved_models/llama2-13b-chat-hf \
    --data-path data/Shepherd/train-alpaca.json \
    --output_dir ./saved_models/lora/${name} \
    --batch_size 8 \
    --micro_batch_size 1 \ 
    --num_epochs 1 \
    --learning_rate 0.0004 \
    --cutoff_len 2048 \
    --val_set_size 0 \
    --lora_r 16 \
    --lora_alpha 16 \
    --lora_dropout 0.05 \
    --lora_target_modules '[gate_proj, down_proj, up_proj]' \
    --train_on_inputs False \
    --add_eos_token False \
    --group_by_length False \
    --prompt_template_name alpaca \
    --lr_scheduler 'cosine' \
    --warmup_steps 100\
    --weight_unlike ${weight_unlike}\
    --threshold ${threshold}\
    --downsample 0.25\

CUDA_VISIBLE_DEVICES=0 python merge.py \
    --base_model_name_or_path saved_models/llama2-13b-chat-hf \
    --peft_model_path ./saved_models/lora/${name} \
    --output_dir ./saved_models/${name}
```

Regarding the full fine-tuning, we did not try full fine-tuning due to the limitations of our computational resources. However, it may be feasible to implement full fine-tuning by eliminating the LORA components in the code and making minor adjustments to fit the model inputs.

Regarding the issue of catastrophic forgetting, our findings presented in Table 5 demonstrate that the model's performance on ARC, HellaSwag, MMLU, and TruthfulQA remains competitive as compared to the base LLM. This suggests that CUT does not compromise the existing knowledge of the base LLM. Consequently, we believe that catastrophic forgetting does not pose a significant challenge in our framework.

**Comment by Labmem009** [comment]
Thank you for your help! 
However, I might have another question: Is the CUT method-trained llama2 model different from the standard llama2 during the inference phase, and is there an inference framework that can be used?

**Comment by Labmem009** [comment]
> Thank you for your help! However, I might have another question: Is the CUT method-trained llama2 model different from the standard llama2 during the inference phase, and is there an inference framework that can be used?


What I mean is that the CUT method mainly alters the loss function during training, so it seems that the trained weights could be used for inference with the standard llama2 functions during the inference phase. Therefore, could an inference framework that is compatible with standard llama2 also be used?

**Comment by wwxu21** [comment]
Yes, your understanding is correct. CUT does not change the model architecture of the base model (i.e. LLaMA2). So you can use the standard llama2 inference function. In short, the way you use CUT-trained models is the same way you use llama2.
In our work and our released model, we use the Alpaca template (system prompt) during training. So we advise utilizing the Alpaca template during inference if employing our released checkpoint.
If you train your customized models with our CUT, we also recommend maintaining consistency in the template for both the training and inference stages.
